neptune/new/internal/backends/operations_preprocessor.py

```diff
@@ -52,17 +52,26 @@
 class OperationsPreprocessor:
     def __init__(self):
         self._accumulators = dict()
+        self.processed_ops_count = 0
 
     def process(self, operations: List[Operation]):
         for op in operations:
-            path_str = path_to_str(op.path)
-            self._accumulators.setdefault(
-                path_str, _OperationsAccumulator(op.path)
-            ).visit(op)
-        result = []
-        for acc in self._accumulators.values():
-            result.extend(acc.get_operations())
-        return result
+            acc: "_OperationsAccumulator" = self._process_op(op)
+            self.processed_ops_count += 1
+            # pylint: disable=protected-access
+            if acc.has_file_ops and len(acc._delete_ops) > 0:
```

Comment:
Since this class started to make some assumptions about in what order returned operations will be performed (returned order is not honored and this class started to be aware of it) following segragating code should be moved here from `execute_oeprations`:
```
upload_operations, artifact_operations, other_operations = [], [], []
for op in operations_preprocessor.get_operations():
    if isinstance(op, (UploadFile, UploadFileContent, UploadFileSet)):
        upload_operations.append(op)
    elif isinstance(op, TrackFilesToArtifact):
        artifact_operations.append(op)
    else:
        other_operations.append(op)
```
and it should have three methods instead of one `get_operations`.

Reply:
sounds great, though I think I'd prefer one method return a dataclass/namedtuple of 3 groups -- retrieving preprocessed operations should still be a single method IMO

otherwise it becomes (at least theoretically) possible to e.g. retrieve upload ops and then pre-process some more operations, which would have weird semantics...